src/reader.rs

```diff
@@ -0,0 +1,136 @@
+use std::fmt::Debug;
+use std::io;
+use std::io::Read;
+
+use endianity::Endianity;
+use leb128;
+use parser::{Error, Result, Format, u64_to_offset};
+
+/// A trait for reading the data from a DWARF section.
+///
+/// All read operations advance the section offset of the reader
+/// unless specified otherwise.
+///
+// TODO: Don't require Copy, Eq
+pub trait Reader: Debug + Clone + Copy + PartialEq + Eq + Read {
+    /// The endianity of bytes that are read.
+    type Endian: Endianity;
```

Comment:
My motivation was that in general, users of `Reader` don't care what the endianity is. It is an implementation detail that should be transparent to them. Using an associated type avoids the noise of declaring the endianity everywhere that `Reader` is used.

I don't think that restricting a type to a single implementation of `Endianity` is a problem. For example, `EndianBuf<LittleEndian>` can only work with `LittleEndian`, not any arbitrary endianity. For types that support dynamic endianity, the implementation of endianity that they use will be `DynamicEndianity` or something like that.